mozmeao/basket
# Add initial CTMS API integration: Populate Email_Id__c in SFDC

Add the initial CTMS API integration, as part of #674.

This change supports this workflow:

* SFDC data is exported to BigQuery (BQ) periodically. Most records have an empty ``Email_Id_c``, the new CTMS ``email_id``.
* BQ adds an ``email_id``, includes this in exports to Acoustic
* CTMS imports from BQ (work in progress)
* Basket fetches from SFDC, and captures ``Email_Id__c`` as ``email_id``. If it is empty, Basket calls CTMS, using a similar interface. If CTMS has a matching record, then Basket uses CTMS's ``email_id``.
* If Basket writes to SFDC to update the record, it populates ``Email_Id__c``, syncing it with BQ.

New settings are used to configure the CTMS backend:
* ``CTMS_ENV`` - 'prod' or 'stage' to select a known ``CTMS_URL``
* ``CTMS_ENABLED`` - True to enable, False to disable CTMS integration
* ``CTMS_URL`` - Select an alternate CTMS API URL
* ``CTMS_CLIENT_ID`` - CTMS Client ID, should start with "id_"
* ``CTMS_CLIENT_SECRET`` - CTMS Client secret, should start with "secret_"

## Manual testing
For local testing, I put my stage credentials in my ``.env`` file, and added some test data, like:

```sh
$ docker-compose run --rm test python manage.py shell
```
```python
from basket.news.backends.ctms import ctms
email = 'test-20210402@example.com'  # you may need to adjust
ctms.interface.post_to_create({'email': {'primary_email': email}})
user_data = ctms.get(email=email)
```

With ``DEBUG=1``, you'll see the CTMS requests flying by. I don't have SFDC credentials to test the full integration.

## Details

``basket.news.backends.ctms`` provides new classes:

* ``CTMSSession`` - Similar to ``requests.session``, with ``session.get``, ``session.post``, etc. It loads an existing token from Redis, and, if there is no token or there is an authentication failure, fetches a token using ``CTMS_CLIENT_ID`` and ``CTMS_CLIENT_SECRET``. There are tests, but with so much mocking that I've considered removing them.
* ``CTMSInferface`` - Provides a method for each of the CTMS API endpoints. Only ``get_by_alternate_id`` and ``get_by_email_id`` are used in current code.
* ``CTMS`` - Provides ``get``, closely modelled on ``SFDC.get``, to load data by the best provided ID and convert it to the flat basket format. It takes a ``CTMSInterface`` to do the work, or ``None`` to silently do nothing.

New functions:

* ``from_vendor``: Converts CTMS nested dictionaries to a Basket dictionary, similar to ``SFDC.from_vendor``
* ``ctms_session``: Creates a ``CTMSSession`` from Django settings
* ``ctms_interface``: Same, for ``CTMSInterface``

New object:
* ``ctms``: An instance of ``CTMS`` with a ``CTMSInterface`` and ``CTMSSession`` based on Django settings, similar to ``sfdc.sfdc``. If CTMS is disabled, ``ctms.interface`` will be ``None``, and ``ctms.get()`` will always return ``None``.

# Tasks
*  ~Add [request-oauthlib](https://requests-oauthlib.readthedocs.io/en/latest/) package~ - Included by pysilverpop
* [x] Add Django settings
* [x] Add low-level interface for authenticated requests
* [x] Add endpoint-specific methods ( ``GET /ctms``, ``GET /ctms/{email_id}``, etc.)
* [x] Read and write SFDC's ``Email_Id__c`` as ``email_id``
* [x] Add ``CTMS.get`` that works in a similar way to ``SFDC.get``, converting to the basket dictionary format
* [x] In ``get_user_data``, call ``CTMS.get`` to populate ``email_id`` when unset in SFDC
* ~Integrate with newsletter subscriptions (send subscriptions to CTMS, CTMS email_id to Salesforce)~ save for future PR
* ~Integrate with preference center (send subscriptions to CTMS, CTMS email_id to Salesforce)~ save for future PR
* [x] Prep for review
